### PR TITLE
[release-1.12][FLINK-20114][connector/kafka,common] Fix a few KafkaSource-related bugs

### PR DESCRIPTION
**Pull request by tweise**
Backport of https://github.com/apache/flink/pull/15161

### CODE REVIEW

**Comment by tweise** [comment]
CC: @lindong28

**Comment by tweise** [comment]
Unrelated e2e test failure:
`2021-05-05T19:51:17.8939009Z May 05 19:51:17 [FAIL] 'Running Kerberized YARN application on Docker test (default input)' failed after 6 minutes and 49 seconds! Test exited with exit code 1`

**Comment by tweise** [comment]
@flinkbot run azure

**Comment by lindong28** [comment]
Thank you @tweise for creating the PR! LGTM.

I have verified that this PR only contains changes from https://github.com/apache/flink/pull/15161. And this PR contains all bug fixes that are in https://github.com/apache/flink/pull/15161 but not currently in the release-1.12 branch.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java
Comment (AHeise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/StreamMultipleInputProcessorFactory.java:343]: Please squash into previous commit.

**Review by AHeise — APPROVED** [review]
LGTM. Let's wait for another approval as we change API in a backport.

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this backport @tweise. The changes look good to me. I couldn't find any API breaking changes. +1 for merging after AZP giving green light.

**Comment by tweise** [comment]
@flinkbot run azure

**Comment by tweise** [comment]
@tillrohrmann @AHeise there seems to be some level of flakiness in the AZP runs and the results are not consistent and contain unrelated failures. What's the recommended way to verify that this PR does not add additional issues? Where can I see the build status of the release-1.12 branch HEAD?

**Comment by AHeise** [comment]
I just deactivated two e2e tests (Kerberos+Yarn) on release-1.12 (couple of hours ago), so it would make sense to rebase. If the Kafka-related (e2e) test pass, it's okay to merge.

**Comment by tweise** [comment]
@AHeise thanks for the suggestion. e2e tests pass after rebase.

`test_ci python` failed in 2 out of 4 AZP runs, that's unrelated to this PR.

Merging..

**Pull request merged**
